src/renderer/pipeline.rs

```diff
@@ -110,7 +108,7 @@ impl<T: RenderEngine> Pipeline<T> {
         io.nav_active = true;
         io.nav_visible = true;
 
-        self.render_loop.before_render(&mut self.ctx);
+        self.render_loop.before_render(&mut self.ctx, &mut self.engine);
```

Comment:
I like this trait based approach.

I originally went for the closure type because I was running into borrow checker issues: over here, `self`, `self.render_loop` and `self.engine` are all mutably borrowed at the same time and the fact that it is compiling at all is kinda throwing me for a loop, but I'm gonna trust Rust on this.

I wonder if `TextureLoader` is the right name or we might just abstract it a little bit further and make it a more generic public `RenderEngine` trait (separated from `crate::renderer::RenderEngine` which should stay private, but also I'd like to find a name which makes it clear that they are distinct and one is meant to be private) so we can add more methods in the future in a non-breaking fashion.

This would be beneficial for covering other future use cases like exposing renderers' internals (to e.g. fetch projection matrices/textures from the game's renderers).

Reply:
“RenderResource”, or "RenderContext/Ctx"？ImguiRenderLoop::render can also have this in param list without issues.

Reply:
Yeah, I'm quite against adding it to `render`. I like the name `RenderContext`, it's just kinda strange to have `RenderContext` in `init` + `before_render` but not in `render`.